src/Silverback.Integration.Kafka/Messaging/Messages/KafkaKeyHelper.cs

```diff
@@ -2,43 +2,86 @@
 // This code is licensed under MIT license (see LICENSE file for details)
 
 using System;
-using System.Globalization;
+using System.Collections.Generic;
 using System.Linq;
-using Silverback.Messaging.HealthChecks;
+using System.Reflection;
+using System.Text;
 
 namespace Silverback.Messaging.Messages
 {
     internal static class KafkaKeyHelper
     {
+        private static readonly Dictionary<Type, PropertyInfo[]> _typeCache = new();
+
         public static string? GetMessageKey(object? message)
         {
             if (message == null)
                 return null;
 
-            if (message is PingMessage)
-                return Guid.NewGuid().ToString("N", CultureInfo.InvariantCulture);
+            PropertyInfo[] propertyInfos = GetPropertyInfos(message);
 
-            var keysDictionary =
-                message.GetType()
-                    .GetProperties()
-                    .Where(propertyInfo => propertyInfo.IsDefined(typeof(KafkaKeyMemberAttribute), true))
-                    .Select(
-                        propertyInfo => new
-                        {
-                            propertyInfo.Name,
-                            Value = propertyInfo.GetValue(message, null)?.ToString()
-                        })
-                    .ToList();
-
-            if (!keysDictionary.Any() ||
-                keysDictionary.All(keyValue => string.IsNullOrEmpty(keyValue.Value)))
+            if (propertyInfos.Length == 0)
             {
                 return null;
             }
 
-            return keysDictionary.Count == 1
-                ? keysDictionary.First().Value
-                : string.Join(",", keysDictionary.Select(p => $"{p.Name}={p.Value}"));
+            if (propertyInfos.Length == 1)
+            {
+                string? value = propertyInfos[0].GetValue(message, null)?.ToString();
+                return string.IsNullOrEmpty(value) ? null : value;
+            }
+
+            StringBuilder stringBuilder = BuildMessageKey(message, propertyInfos);
+            return stringBuilder.Length == 0 ? null : stringBuilder.ToString();
+        }
+
+        private static PropertyInfo[] GetPropertyInfos(object message)
+        {
+            if (_typeCache.TryGetValue(message.GetType(), out PropertyInfo[] propertyInfos))
+            {
+                return propertyInfos;
+            }
+
+            lock (_typeCache)
```

Comment:
Why don't you use a `ConcurrentDictionary`? I believe it would be _much better_ in this case. It's useless to sync lock all producers, producing different types.

The code would also be more concise.

Why did you chose this path?

Reply:
Because before .NET 5 (I just learned that this isn't the case anymore for .NET 5), the concurrent dict looked on read as well and reading was slower than with a dict. Because the lock on write would be enough for this use case, I did it like this. 
However, as concurrentDict read is even faster in .NET 5 than the read on the normal dict, I will change it. Ok?

Reply:
I see and I actually think I knew that, just never took it into consideration. If it was fixed in .NET 5 then yes, switch to a `ConcurrentDictionary`.

Reply:
Because before .NET 5 (I just learned that this isn't the case anymore for .NET 5), the concurrent dict looked on read as well and reading was slower than with a dict.// |                          Method |      Mean |     Error |    StdDev |   Gen 0 | Gen 1 | Gen 2 | Allocated |
// |-------------------------------- |----------:|----------:|----------:|--------:|------:|------:|----------:|
// |           MessagesWithSimpleKey | 14.907 us | 0.3692 us | 1.0887 us |       - |     - |     - |         - |
// | MessagesWithSimpleKeyButNoValue | 14.219 us | 0.2720 us | 0.7353 us |       - |     - |     - |         - |
// |         MessagesWithCombinedKey | 73.377 us | 1.4529 us | 2.3461 us | 20.9961 |     - |     - |  44,000 B |
// |              MessagesWithoutKey |  3.248 us | 0.0649 us | 0.0866 us |       - |     - |     - |         - |

// |                          Method |      Mean |     Error |    StdDev |   Gen 0 | Gen 1 | Gen 2 | Allocated |
// |-------------------------------- |----------:|----------:|----------:|--------:|------:|------:|----------:|
// |           MessagesWithSimpleKey | 13.955 us | 0.4092 us | 1.1674 us |       - |     - |     - |         - |
// | MessagesWithSimpleKeyButNoValue | 13.471 us | 0.2694 us | 0.7466 us |       - |     - |     - |         - |
// |         MessagesWithCombinedKey | 75.546 us | 1.6008 us | 4.6186 us | 20.9961 |     - |     - |  44,000 B |
// |              MessagesWithoutKey |  2.860 us | 0.0712 us | 0.2043 us |       - |     - |     - |         - |